internal/sourcemap/sourcemapper.go

```diff
@@ -0,0 +1,211 @@
+}
+
+func (sm *sourceMapper) ClearCache() {
+	sm.documentPositionMappers = make(map[string]DocumentPositionMapper)
+}
+
+func (sm *sourceMapper) getDocumentPositionMapper(generatedFileName string, sourceFileName string) DocumentPositionMapper {
+	path := sm.toPath(generatedFileName)
+	if mapper, exists := sm.documentPositionMappers[path]; exists {
+		return mapper
+	}
+
+	mapper := sm.createDocumentPositionMapper(generatedFileName, sourceFileName)
+	sm.documentPositionMappers[path] = mapper
+	return mapper
+}
+
+func (sm *sourceMapper) createDocumentPositionMapper(generatedFileName string, sourceFileName string) DocumentPositionMapper {
+	// First try to read the generated file to get source mapping URL
+	content, ok := sm.fileReader.ReadFile(generatedFileName)
+	if !ok {
+		return IdentityDocumentPositionMapper()
+	}
+
+	mapURL := tryGetSourceMappingURL(content)
+	if mapURL == "" {
+		return IdentityDocumentPositionMapper()
+	}
+
+	var mapFileName string
+	if strings.HasPrefix(mapURL, "/") {
+		mapFileName = mapURL
+	} else {
+		dir := tspath.GetDirectoryPath(generatedFileName)
+		mapFileName = tspath.CombinePaths(dir, mapURL)
+	}
+
+	mapContent, ok := sm.fileReader.ReadFile(mapFileName)
+	if !ok {
+		return IdentityDocumentPositionMapper()
+	}
+
+	return CreateDocumentPositionMapper(sm, mapContent, mapFileName)
+}
+
+func (sm *sourceMapper) toPath(fileName string) string {
+	return string(tspath.ToPath(fileName, sm.currentDirectory, sm.host.UseCaseSensitiveFileNames()))
+}
+
+// DocumentPositionMapperHost implementation
+func (sm *sourceMapper) GetSource(fileName string) Source {
+	content, ok := sm.fileReader.ReadFile(fileName)
+	if !ok {
+		return nil
+	}
+
+	return NewSimpleSourceFile(fileName, content)
+}
+
+func (sm *sourceMapper) GetCanonicalFileName(path string) string {
+	return sm.getCanonicalFileName(path)
+}
+
+func (sm *sourceMapper) Log(text string) {
+	// Could be implemented to use host's logging if needed
+	// For now, we'll keep it simple
+}
+
+// SimpleSourceFile implements Source interface for raw text content
+type SimpleSourceFile struct {
+	fileName string
+	text     string
+	lineMap  []core.TextPos
+}
+
+func (sf *SimpleSourceFile) FileName() string {
+	return sf.fileName
+}
+
+func (sf *SimpleSourceFile) Text() string {
+	return sf.text
+}
+
+func (sf *SimpleSourceFile) LineMap() []core.TextPos {
+	return sf.lineMap
+}
+
+// NewSimpleSourceFile creates a SimpleSourceFile from fileName and text content
+func NewSimpleSourceFile(fileName, text string) *SimpleSourceFile {
+	return &SimpleSourceFile{
+		fileName: fileName,
+		text:     text,
+		lineMap:  core.ComputeLineStarts(text),
+	}
+}
+
+// Helper functions
+
+func isDeclarationFileName(fileName string) bool {
+	return strings.HasSuffix(fileName, ".d.ts")
+}
+
+func createGetCanonicalFileName(useCaseSensitiveFileNames bool) func(string) string {
+	if useCaseSensitiveFileNames {
+		return func(path string) string { return path }
+	}
+	return strings.ToLower
+}
```

Comment:
This isn't a correct implementation, unfortunately. If we need to compare paths like this, we use `tspath.ToPath`.

We also don't pass around "canonical filename functions" anymore, instead choosing to pass around Paths and the case sensitivity.